### PR TITLE
rados: refuse to cppool if there are snaps; warn about user_version

### PR DESCRIPTION
**Pull request by liewegas**
Alternative to https://github.com/ceph/ceph/pull/8975

### CODE REVIEW

**Comment by liewegas** [comment]
@jdurgin @yehudasa what do you think?

File: src/tools/rados/rados.cc [hunk 1]
```diff
@@ -2616,6 +2617,15 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
       goto out;
     }
 
+    if (rados.get_pool_is_selfmanaged_snaps_mode(src_pool)) {
+      cerr << "WARNING: pool " << src_pool << " has selfmanaged snaps, which are not preserved\n"
+	   << "    by the cppool operation.  This will break any snapshot user.\n"
+	   << "    If you insist on making a broken copy, you can pass\n"
+	   << "    --yes-i-really-mean-it to proceed anyway."
+	   << std::endl;
+      if (!force)
```
Comment (yehudasa) [inline src/tools/rados/rados.cc]: @liewegas maybe split the cerr message, and have the second part (starting at 'If you insist') dumped only if !force?

**Comment by yehudasa** [comment]
@liewegas looks fine (see minor comment)

**Comment by liewegas** [comment]
updated

**Comment by liewegas** [comment]
@athanatos if we just add a set_user_version rados op, we can resolve that problem with cppool as well and remove the warning.  Is there a reason not to add it?  We already have a get_user_version op.

It can be ENOTSUP on ec overwrite pools, I guess...

**Comment by jdurgin** [comment]
the warnings lgtm

**Comment by athanatos** [comment]
Using set_user_version would be fine, didn't know it existed.

**Comment by liewegas** [comment]
@athanatos it doesn't, but we could add it...

**Comment by yehudasa** [comment]
@liewegas @athanatos  but then we could have multiple objects with the same user version. Or with user version that will be bigger than the max user version for the target pool. The latter might be an issue with radosgw (not 100% sure it is).

**Comment by liewegas** [comment]
The pool max automatically bumps up so it will always be greater, but cache proxy writes mean you could get the same version on different objects.  Is that a problem for RGW?

**Comment by liewegas** [comment]
2016-05-23T09:56:34.461 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [----------] 5 tests from LibRadosSnapshotsPP
2016-05-23T09:56:34.461 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [ RUN      ] LibRadosSnapshotsPP.SnapListPP
2016-05-23T09:56:34.461 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: test/librados/snapshots.cc:46: Failure
2016-05-23T09:56:34.462 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: Value of: cluster.get_pool_is_selfmanaged_snaps_mode(pool_name)
2016-05-23T09:56:34.462 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots:   Actual: false
2016-05-23T09:56:34.462 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: Expected: true
2016-05-23T09:56:34.462 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [  FAILED  ] LibRadosSnapshotsPP.SnapListPP (2998 ms)
2016-05-23T09:56:34.462 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [ RUN      ] LibRadosSnapshotsPP.SnapRemovePP
2016-05-23T09:56:34.462 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: test/librados/snapshots.cc:75: Failure
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: Value of: ioctx.snap_create("snap1")
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots:   Actual: -17
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: Expected: 0
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [  FAILED  ] LibRadosSnapshotsPP.SnapRemovePP (15 ms)
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [ RUN      ] LibRadosSnapshotsPP.RollbackPP
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: test/librados/snapshots.cc:103: Failure
2016-05-23T09:56:34.463 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: Value of: ioctx.snap_create("snap1")
2016-05-23T09:56:34.464 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots:   Actual: -17
2016-05-23T09:56:34.464 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: Expected: 0
2016-05-23T09:56:34.464 INFO:tasks.workunit.client.0.smithi005.stdout:            api_snapshots: [  FAILED  ] LibRadosSnapshotsPP.RollbackPP (11 ms)

**Comment by liewegas** [comment]
/a/sage-2016-05-23_05:47:51-rados-wip-sage-testing---basic-smithi/208709

**Comment by liewegas** [comment]
fixed test

**Comment by yuriw** [comment]
tested http://pulpito.ceph.com/yuriw-2016-05-25_20:42:59-rados-wip-yuri-testing-distro-basic-smith

**Pull request merged**